File: frontend/src/types/notification.ts
// types/notification.ts

/**
 * 알림 종류.
 * - nudge: AI가 먼저 보낸 독촉 채팅 (일반 채팅은 알림을 만들지 않는다)
 * - todoAdded / todoDone: 투두가 추가되거나 완료 체크됨
 * - todoIncomplete: 밤 11시까지 완료하지 않은 투두가 남아 있음
 * - plannerIncomplete: 밤 11시까지 텐미닛 플래너가 비어 있음
 */
export type NotificationType =
  'nudge' | 'todoAdded' | 'todoDone' | 'todoIncomplete' | 'plannerIncomplete';

/** 홈 헤더의 알림 목록에 뜨는 알림 하나 */
export interface AppNotification {
  id: string;
  type: NotificationType;
  /** 위쪽 작은 제목. AI 이름이나 목표 이름이 들어간다 */
  title: string;
  /** 아래 본문 */
  body: string;
  /** 발생 시각 (ISO 8601 문자열) */
  createdAt: string;
  isRead: boolean;
  /** 눌렀을 때 이동할 경로 (예: 해당 채팅방) */
  linkTo?: string;
}

/** 목록에 유지하는 최대 알림 개수. 넘치면 오래된 것부터 사라진다 */
export const MAX_NOTIFICATIONS = 5;

File: backend/app/models/notification.py
"""알림 모델 (api.md §5.2).

type: nudge | todoAdded | todoDone | todoIncomplete | plannerIncomplete
- nudge/todoAdded/todoDone → 해당 채팅방으로 이동
- todoIncomplete/plannerIncomplete → 기록 탭으로 이동

link_to: 프론트가 바로 이동할 경로 문자열(예: /chat/g_01H, /record).
(경로 조립을 서버가 하는 방식 — 프론트 AppNotification.linkTo 계약에 맞춤)
"""

from datetime import UTC, datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base
from app.core.ids import new_id

NOTIFICATION_TYPES = (
    "nudge",
    "todoAdded",
    "todoDone",
    "todoIncomplete",
    "plannerIncomplete",
)


def _now() -> datetime:
    return datetime.now(UTC)


class Notification(Base):
    __tablename__ = "notifications"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: new_id("n"))
    user_id: Mapped[str] = mapped_column(
        String, ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    type: Mapped[str] = mapped_column(String, nullable=False)
    title: Mapped[str] = mapped_column(String, nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False, default="")
    link_to: Mapped[str | None] = mapped_column(String, nullable=True)
    is_read: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now, index=True)
